controllers/rollout/trafficrouting.go

```diff
@@ -0,0 +1,283 @@
+	return true, nil
+}
+
+func (r *rolloutContext) restoreStableService() (bool, error) {
+	//fetch stable service
+	sName := r.rollout.Spec.Strategy.Canary.TrafficRouting.Service
+	r.stableService = &corev1.Service{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: r.rollout.Namespace, Name: sName}, r.stableService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		klog.Errorf("rollout(%s/%s) get stable service(%s) failed: %s", r.rollout.Namespace, r.rollout.Name, sName, err.Error())
+		return false, err
+	}
+
+	if r.newStatus.CanaryStatus == nil {
+		r.newStatus.CanaryStatus = &appsv1alpha1.CanaryStatus{}
+	}
+	//restore stable service configuration，remove hash revision selector
+	if r.stableService.Spec.Selector != nil && r.stableService.Spec.Selector[r.podRevisionLabelKey()] != "" {
+		delete(r.stableService.Spec.Selector, r.podRevisionLabelKey())
+		if err := r.retryUpdateService(r.stableService); err != nil {
+			return false, err
+		}
+		klog.Infof("remove rollout(%s/%s) stable service(%s) pod revision selector success, and retry later", r.rollout.Namespace, r.rollout.Name, r.stableService.Name)
+		r.newStatus.CanaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+		return false, nil
+	}
+	// After restore stable service configuration, give the ingress provider 3 seconds to take effect
+	if r.newStatus.CanaryStatus.LastUpdateTime != nil {
+		if verifyTime := r.newStatus.CanaryStatus.LastUpdateTime.Add(time.Second * 3); verifyTime.After(time.Now()) {
+			klog.Infof("restore rollout(%s/%s) stable service(%s) done, and wait a moment", r.rollout.Namespace, r.rollout.Name, r.stableService.Name)
+			return false, nil
+		}
+	}
+	klog.Infof("rollout(%s/%s) doFinalising restore stable service(%s) success", r.rollout.Namespace, r.rollout.Name, r.stableService.Name)
+	return true, nil
+}
+
+func (r *rolloutContext) doFinalisingTrafficRouting() (bool, error) {
+	if r.newStatus.CanaryStatus == nil {
+		r.newStatus.CanaryStatus = &appsv1alpha1.CanaryStatus{}
+	}
+
+	// 1. restore ingress and route traffic to stable service
+	trController, err := r.newTrafficRoutingController(r)
+	if err != nil {
+		klog.Errorf("rollout(%s/%s) newTrafficRoutingController failed: %s", r.rollout.Namespace, r.rollout.Name, err.Error())
+		return false, err
+	}
+	verify, err := trController.VerifyTrafficRouting(-1)
+	if err != nil {
+		return false, err
+	} else if !verify {
+		r.newStatus.CanaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+		err = trController.SetRoutes(0)
+		if err != nil && errors.IsNotFound(err) {
+			klog.Warningf("rollout(%s/%s) VerifyTrafficRouting(-1), and stable ingress not found", r.rollout.Namespace, r.rollout.Name)
+			return false, nil
+		}
+		return false, err
+	}
+
+	// After do TrafficRouting configuration, give the ingress provider 5 seconds to take effect
```

Comment:
I am a little confused about this time 5 seconds, is this only about nginx ingress? what if other ingress provider?

Reply:
Ditto, later also intend to add hook mechanism, to check